InnerTypeLastCheck: false negative on compact source files

Parent Issue #20590

I have read check documentation: https://checkstyle.org/checks/design/innertypelast.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="InnerTypeLast"/>
  </module>
</module>
```

```java
$ cat src/InputInnerTypeLastCompact.java
class Inner { }

void main() { }
```

```
$ java  -jar checkstyle-13.8.0-all.jar -c config.xml src/InputInnerTypeLastCompact.java
Starting audit...
Audit done.
```

**Describe what you expect in detail.**
No violation is reported but per the check own documented example a member
declared after an inner type should violate

```java
$ cat src/Test.java
class Test1 {
  private String s;
  class InnerTest1 {}
  public void test() {} // violation
}

$ java  -jar checkstyle-13.8.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] C:\Users\PMYLS\Desktop\check\src\Test.java:4:3: Init blocks, constructors, fields and methods should be before inner types. [InnerTypeLast]
Audit done.
Checkstyle ends with 1 errors.

```

This is confirmed by wrapping the same code in an ordinary named class which
correctly reports the violation

```java
$ cat src/InputInnerTypeLastCompact.java
public class InputInnerTypeLastCompact {
    class Inner { }

    void main() { }
}
```

```
$ java  -jar checkstyle-13.8.0-all.jar -c config.xml src/InputInnerTypeLastCompact.java
Starting audit...
[ERROR] C:\Users\PMYLS\Desktop\check\src\InputInnerTypeLastCompact.java:4:5: Init blocks, constructors, fields and methods should be before inner types. [InnerTypeLast]
Audit done.
Checkstyle ends with 1 errors.
```


**Side effect of fix:**
now content of Enums are also validated.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/InnerTypeLastCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/InnerTypeLastCheck.java
@@ -19,14 +19,13 @@
 
 package com.puppycrawl.tools.checkstyle.checks.design;
 
-import java.util.BitSet;
+import java.util.Set;
 
-import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
 import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
-import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
  * <div>
@@ -37,7 +36,7 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  *
  * @since 5.2
  */
-@FileStatefulCheck
+@StatelessCheck
 public class InnerTypeLastCheck extends AbstractCheck {
 
     /**
@@ -47,7 +46,7 @@ public class InnerTypeLastCheck extends AbstractCheck {
     public static final String MSG_KEY = "arrangement.members.before.inner";
 
     /** Set of class member tokens. */
-    private static final BitSet CLASS_MEMBER_TOKENS = TokenUtil.asBitSet(
+    private static final Set<Integer> CLASS_MEMBER_TOKENS = Set.of(
             TokenTypes.VARIABLE_DEF,
             TokenTypes.METHOD_DEF,
             TokenTypes.CTOR_DEF,
@@ -55,9 +54,6 @@ public class InnerTypeLastCheck extends AbstractCheck {
             TokenTypes.STATIC_INIT,
             TokenTypes.COMPACT_CTOR_DEF
     );
-
-    /** Meet a root class. */
-    private boolean rootClass;
 
     @Override
     public int[] getDefaultTokens() {
@@ -79,32 +75,14 @@ public class InnerTypeLastCheck extends AbstractCheck {
     }
 
     @Override
-    public void beginTree(DetailAST rootAST) {
-        rootClass = true;
-    }
-
-    @Override
     public void visitToken(DetailAST ast) {
-        // First root class
-        if (rootClass) {
-            rootClass = false;
-        }
-        else {
-            DetailAST nextSibling = ast;
-            while (nextSibling != null) {
-                if (!ScopeUtil.isInCodeBlock(ast)
-                        && CLASS_MEMBER_TOKENS.get(nextSibling.getType())) {
-                    log(nextSibling, MSG_KEY);
-                }
-                nextSibling = nextSibling.getNextSibling();
+        DetailAST nextSibling = ast;
+        while (nextSibling != null) {
+            if (!ScopeUtil.isInCodeBlock(ast)
+                    && CLASS_MEMBER_TOKENS.contains(nextSibling.getType())) {
+                log(nextSibling, MSG_KEY);
             }
-        }
-    }
-
-    @Override
-    public void leaveToken(DetailAST ast) {
-        if (TokenUtil.isRootNode(ast.getParent())) {
-            rootClass = true;
+            nextSibling = nextSibling.getNextSibling();
         }
     }
 

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/design/InnerTypeLastCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/design/InnerTypeLastCheckTest.java
@@ -29,7 +29,6 @@ import org.junit.jupiter.api.Test;
 import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
 import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
-import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
 
 public class InnerTypeLastCheckTest extends AbstractModuleTestSupport {
 
@@ -66,14 +65,19 @@ public class InnerTypeLastCheckTest extends AbstractModuleTestSupport {
 
     @Test
     public void testIfRootClassChecked() throws Exception {
-        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        final String[] expected = {
+            "20:5: " + getCheckMessage(MSG_KEY),
+            "23:5: " + getCheckMessage(MSG_KEY),
+        };
         verifyWithInlineConfigParser(
                 getPath("InputInnerTypeLastClassRootClass.java"), expected);
     }
 
     @Test
     public void testIfRootClassChecked2() throws Exception {
-        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        final String[] expected = {
+            "25:5: " + getCheckMessage(MSG_KEY),
+        };
         verifyWithInlineConfigParser(
                 getPath("InputInnerTypeLastClassRootClass2.java"), expected);
     }
@@ -82,11 +86,16 @@ public class InnerTypeLastCheckTest extends AbstractModuleTestSupport {
     public void testIfRootClassChecked3() throws Exception {
         final DefaultConfiguration checkConfig =
             createModuleConfig(InnerTypeLastCheck.class);
-        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        final String[] expected = {
+            "20:5: " + getCheckMessage(MSG_KEY),
+            "23:5: " + getCheckMessage(MSG_KEY),
+            "20:5: " + getCheckMessage(MSG_KEY),
+            "23:5: " + getCheckMessage(MSG_KEY),
+        };
         verify(createChecker(checkConfig), new File[] {
             new File(getPath("InputInnerTypeLastClassRootClass.java")),
             new File(getPath("InputInnerTypeLastClassRootClass.java")),
-        }, "InputInnerTypeLastClassRootClass.java", expected);
+        }, getPath("InputInnerTypeLastClassRootClass.java"), expected);
     }
 
     @Test
@@ -141,4 +150,22 @@ public class InnerTypeLastCheckTest extends AbstractModuleTestSupport {
             .isEqualTo(expected);
     }
 
+    @Test
+    public void testInnerTypeLastCompactSourceFile() throws Exception {
+        final String[] expected = {
+            "12:1: " + getCheckMessage(MSG_KEY),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("compact/InputInnerTypeLastCompactSourceFile.java"), expected);
+    }
+
+    @Test
+    public void testInnerTypeLastCompactSourceFileNested() throws Exception {
+        final String[] expected = {
+            "15:5: " + getCheckMessage(MSG_KEY),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("compact/InputInnerTypeLastCompactSourceFileNested.java"),
+                expected);
+    }
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksCompactSourceCoverageTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksCompactSourceCoverageTest.java
@@ -149,7 +149,6 @@ public class AllChecksCompactSourceCoverageTest {
         "ImportOrderCheck",
         "IndentationCheck",
         "InnerAssignmentCheck",
-        "InnerTypeLastCheck",
         "InterfaceIsTypeCheck",
         "InterfaceMemberImpliedModifierCheck",
         "InterfaceTypeParameterNameCheck",

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/compact/InputInnerTypeLastCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/compact/InputInnerTypeLastCompactSourceFile.java
@@ -1,0 +1,12 @@
+/*
+InnerTypeLast
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+class Inner { }
+
+// violation below 'Init blocks, constructors, fields and methods should be before inner types'
+void main() { }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/compact/InputInnerTypeLastCompactSourceFileNested.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/compact/InputInnerTypeLastCompactSourceFileNested.java
@@ -1,0 +1,16 @@
+/*
+InnerTypeLast
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+void main() { }
+
+class Outer {
+    class Nested { }
+
+    // violation below 'Init blocks, constructors, fields and methods should be before inner types'
+    void process() { }
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/InputInnerTypeLastClassRootClass.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/InputInnerTypeLastClassRootClass.java
@@ -16,8 +16,10 @@ public enum InputInnerTypeLastClassRootClass {
         public static final int NO = 4;
     }
 
+    // violation below 'Init blocks, constructors, fields and methods should be before inner types.'
     private final int bits;
 
+    // violation below 'Init blocks, constructors, fields and methods should be before inner types.'
     private InputInnerTypeLastClassRootClass(int bits) {
         this.bits = bits;
     }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/InputInnerTypeLastClassRootClass2.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/innertypelast/InputInnerTypeLastClassRootClass2.java
@@ -21,6 +21,7 @@ enum InputInnerTypeLastClassRootClass2a
     {
     }
 
+    // violation below 'Init blocks, constructors, fields and methods should be before inner types.'
     private int mFoo = 0;
 
     class AsyncProcess {

